### PR TITLE
KEYCLOAK-12243 MultivaluedHashMap.equalsIgnoreValueOrder() returns true for some maps that are not equal

### PR DESCRIPTION
**Pull request by tkyjovsk**
fix + test

### CODE REVIEW

**Comment by toziol** [comment]
You should add this test :

        map.clear();
        map.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
        differentMap.clear();
        differentMap.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
        differentMap.put(2, Arrays.asList(new Integer[]{4, 5, 6}));
        equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(map, differentMap); // diff entrySet size

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java
Comment (hmlnarik) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Why `final`?

Comment (tkyjovsk) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Method used in the constructor shouldn't be overridable.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (fperot74) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: This is not false but I'm mitigated about this test : map was first initialized like if it was a MultivaluedHashMap<Integer, Integer>...
Here it is used like if it was a MultivaluedHashMap<String, String>. Mixing types is not what the object has been created for but it is not considered as an error...
Ideally, MultivaluedHashMap should not have mixed type content and types should be provided at declaration... but here, it is just for test purpose.

Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Yes, it was just for the test. But you're right, the types shouldn't be mixed. I addressed your suggestions in the next push. Thanks!

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Resolved in `4e01018`.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Thanks for this suggestion. Resolved in `4e01018`.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (tkyjovsk) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: This needs to be fixed to:
```
        equalMap.put(1, Arrays.asList(4, 3, 2, 5, 1));
        equalMap.put(2, Arrays.asList(6, 7, 4, 8, 5));
        equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(map, equalMap);
```

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java
Comment (hmlnarik) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Thanks for explanation. I did not look around the PR.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (fperot74) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: You can add <T, R> in both equalsIgnoreValueOrder_XXX methods like this:
private <T, R> void equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(MultivaluedHashMap<T, R> map, MultivaluedHashMap<T, R> equalMap) {

**Review by fperot74 — DISMISSED** [review]

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: I added the generics, thanks.

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java
Comment (mposolda) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: This is in the "common" module, which is used by adapters etc and require to compile on some historic java versions (JDK 7 etc). Due this, I would not use `ArrayList<>` , but rather keep this line as  `ArrayList<V>`  . As it was before

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java
Comment (mposolda) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Same as above

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (mposolda) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Same as above

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java
Comment (mposolda) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Same as above

**Review by mposolda — CHANGES_REQUESTED** [review]
I've added few minor comments, but all of them are for the same thing. If it is possible, I would avoid using construct like `List<V> l = new ArrayList<>();` or `Map<K,V> m = new HashMap<>()` , but keep it everywhere in the old format to ensure it can safely compile on historic java versions. That applies to test as well.

Besides that, it looks good and I would approve once @fperot74 approves.

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java
Comment (tkyjovsk) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Java 7 should support type inference for generics. I tried compiling the module with JDK 7 and it compiled fine. (There was some problem with surefire plugin but that seems unrelated.)

Comment (mposolda) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Cool, Thanks! I don't know why I though it doesn't :) Approving and merging then.

**Review by mposolda — APPROVED** [review]

**Pull request merged**
